enhancements/installer/aws-custom-dns.md

```diff
@@ -0,0 +1,488 @@
+configuration of cloud DNS while continuing with LB configuration.
+
+6. The Ingress Controllers managed by the cluster-ingress-operator continue
+to configure the application LBs to enable in-cluster and external accesss to
+cluster services. The Ingress Controllers will continue to update the DNSRecord
+ConfigResource(CR) with the configuration required to configure DNS for these
+services.
+
+7. The MCO will watch the DNSRecord CR for the DNS configuration mentioned in
+step 6. MCO will update the newly added `IngressDNSConfig` field within the
+`AWSPlatformStatus` with the DNS configuartion for *.apps records by reading
+the DNSRecord CR.
+
+8. The runtimecfg utility will monitor the `Infrastructure` CR for changes to
+`IngressDNSConfig` of `AWSPlatformStatus` to update the CoreDNS pod's
+CoreFile to now also include DNS configuration for both internal and
+external access to the cluster services (the `*.apps` records).
+
+9. Once the cluster is functional, the user can look at the contents of the new
+fields `APIServerDNSConfig` and `IngressDNSConfig` in the `AWSPlatformStatus`
+field of the `Infrastructure` CR to configure their external DNS with entries
+for `api` and `*.apps` service.
+
+### API Extensions
+
+1. The AWSPlatformStatus within the PlatformStatus field of the Infrastructure
+ConfigResource (CR) is updated to contain DNS config required to configure
+the in-cluster CoreDNS solution. This same CR is available to the user post
+a successful cluster install, to configure their own DNS solution.
+
+```go
+type AWSPlatformStatus struct {
+        <snip>
+        // APIServerDNSConfig contains information to configure DNS for API Server.
+        // This field will be set only when the userConfiguredDNS feature is enabled.
+        APIServerDNSConfig []DNSConfig `json:"apiServerDNSConfig,omitempty"`
+
+        // InternalAPIServerDNSConfig contains information to configure DNS for the Internal API Server.
+        // This field will be set only when the userConfiguredDNS feature is enabled.
+        InternalAPIServerDNSConfig []DNSConfig `json:"InternalAPIServerDNSConfig,omitempty"`
+
+        // IngressDNSConfig contains information to configure DNS for cluster services.
+        // This field will be set only when the userConfiguredDNS feature is enabled.
+        IngressDNSConfig []DNSConfig `json:"ingressDNSConfig,omitempty"`
+}
+
+type DNSConfig struct {
+       RecordType  string `json:"recordType"`
+       LBIPAddress string `json:"lbIPAddress"`
+}
+
+```
+
+2. Install config is updated to allow the customer to specify if an external
+user configured DNS will be used. `UserConfiguredDNS` is added to the
+install-config and will have to be explicitly set to `Enabled` to enable this
+functionality. This config is not added to any platform specific section of
+the config because there are plans to allow this functionality in Azure and GCP
+too. The validation for this config will disallow this value being `Enabled` in
```

Comment:
So this option will be universal but doesn't apply to platforms other than AWS, Azure and GCP? Better IMO to add the option to each platform's configuration else you end up with configuration that a user could set, that then doesn't apply. It may not be obvious that this option doesn't apply to say, BareMetal platforms if it's at a parent scope. Within the specific platform configuration it's much more obvious

Reply:
There is precedent for both options. Here is an example of config added outside platform specific sections. https://github.com/openshift/installer/blob/dbcb9b2930b985e0c35d8e55c8bf96be86629428/pkg/types/installconfig.go#L161. The validation code makes sure to inform the user when it is configured with an incorrect value on the wrong platform. CredentialsMode is another example. 
I see this config as an attribute of the installation and not of the underlying cloud so, I see this being in the non-platform specific section.

Reply:
I would go towards making this in platform specific as it's more obvious which platforms it's supported in, just from the schema, rather than being validated later, but if you feel strongly, since this is in install config, not much I can do about it ;)

Reply:
I don't have a strong opinion on whether the field should be top-level or a child of platform. I think Joel's suggestion of having it as a child of platform makes sense.